platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,826 @@
+| 7 | Pre-Genesis: Separate app with genesis roster creates key material in state before network officially starts. | <span style="color:green">YES</span> | <span style="color:green">YES</span>                 | <span style="color:green">YES</span> | <span style="color:green">YES</span> | <span style="color:red">NO</span>    | <span style="color:red">NO</span>    | <span style="color:red">NO</span>    | <span style="color:red">HIGH</span>       | Applies to new networks only.  Use Option 1 or 2 for existing networks.            |
+| 8 | Instead of separate app, detect genesis, key the state, network restart from round 0 with keyed state.        | <span style="color:green">YES</span> | <span style="color:green">YES</span>                 | <span style="color:green">YES</span> | <span style="color:green">YES</span> | <span style="color:red">NO</span>    | <span style="color:red">NO</span>    | <span style="color:green">YES</span> | <span style="color:red">HIGH</span>       | Applies to new networks only.  Use Option 1 or 2 for existing networks.            | 
+
+Option 2 was chosen for existing networks and Option 8 was selected for new networks. This decision was made to
+ensure that the first block in the block stream is signed by the ledger and that new networks always have block 0
+signed.
+
+---
+
+## Changes
+
+The changes are presented in the following order:
+
+1. Public API
+2. Core Behaviors
+3. Component Architecture
+4. Configuration
+5. Metrics
+6. Performance
+
+### Public API
+
+The public API changes consist of the following:
+
+* The Ledger Signature API
+* System Transactions
+* State Data Structures
+
+#### Ledger Signature API
+
+The Ledger Signature API is the interface between the application and the platform by which the application submits
+a message to be signed by the ledger, the nodes sign the message with their shares, and the platform aggregates the
+signatures into a ledger signature. The ledger signature is then asynchronously returned to the application.
+
+To request a ledger signature, the application invokes a new method on the `State`:
+
+* `State::sign(byte[] message) throws NotReadyException`
+
+To receive a ledger signature, the application registers a callback with the platform builder:
+
+* `PlatformBuilder::withLedgerSignatureCallback(Consumer<PairingSignature> ledgerSignatureConsumer)`
+
+#### System Transactions
+
+The following are new system transactions needed to key rosters and generate ledger signatures.
+
+##### TssMessage System Transaction
+
+TODO: Protobuf of data
+
+##### TssVote System Transaction
+
+TODO: Protobuf of data
+
+##### TssShareSignature System Transaction
+
+TODO: Protobuf of data
+
+#### State Datastructures
+
+The state needs to store the relevant ledger id, consensus rosters, key material, and partially constructed ledger
+signatures.
+
+##### Ledger Id
+
+The `ledgerId` is the public ledger key able to verify ledger signatures. It is used by both the application and the
```

Comment:
What use does the app have for the ledger id?

Reply:
The app, perhaps no use.  but the block node would have a use.

Reply:
Modified the text.